src/plan.ts

```diff
@@ -0,0 +1,186 @@
+    ) as object
+    // @ts-ignore
+    let name: string
+    if ('name' in charmcraft) {
+      name = charmcraft['name'] as string
+    } else {
+      const metadataFile = path.join(
+        path.dirname(charmcraftFile),
+        'metadata.yaml'
+      )
+      const metadata = yaml.load(
+        fs.readFileSync(metadataFile, { encoding: 'utf-8' })
+      ) as object
+      if (!('name' in metadata)) {
+        throw new Error(`unknown charm name (${workingDir})`)
+      }
+      name = metadata['name'] as string
+    }
+    return {
+      type: 'charm',
+      name,
+      source_file: file,
+      source_directory: path.dirname(file),
+      output_type: 'file',
+      output: sanitizeArtifactName(`${id}__build__output__charm__${name}`)
+    }
+  })
+}
+
+async function planBuildRock(
+  workingDir: string,
+  id: string,
+  outputType: 'file' | 'registry'
+): Promise<BuildPlan[]> {
+  const rockcraftFiles = await (
+    await glob.create(path.join(workingDir, '**', 'rockcraft.yaml'))
+  ).glob()
+  return rockcraftFiles.map((rockcraftFile: string) => {
+    const file = path.join(workingDir, path.relative(workingDir, rockcraftFile))
+    const rockcraft = yaml.load(
+      fs.readFileSync(rockcraftFile, { encoding: 'utf-8' })
+    )
+    // @ts-ignore
+    const name = rockcraft['name']
+    return {
+      type: 'rock',
+      name,
+      source_file: file,
+      source_directory: path.dirname(file),
+      output_type: outputType,
+      output: sanitizeArtifactName(`${id}__build__output__rock__${name}`)
+    }
+  })
+}
+
+async function planBuildDockerImage(
+  workingDir: string,
+  id: string,
+  outputType: 'file' | 'registry'
+): Promise<BuildPlan[]> {
+  const dockerFiles = await (
+    await glob.create(path.join(workingDir, '**', '*.Dockerfile'))
+  ).glob()
+  return dockerFiles.map((dockerFile: string) => {
+    const file = path.join(workingDir, path.relative(workingDir, dockerFile))
+    const name = path.basename(file).replace(/.Dockerfile$/, '')
+    return {
+      type: 'docker-image',
+      name,
+      source_file: file,
+      source_directory: path.dirname(file),
+      output_type: outputType,
+      output: sanitizeArtifactName(
+        `${id}__build__output__docker-image__${name}`
+      )
+    }
+  })
+}
+
+async function planBuild(
+  workingDir: string,
+  id: string,
+  imageOutputType: 'file' | 'registry'
+): Promise<BuildPlan[]> {
+  return [
+    ...(await planBuildCharm(workingDir, id)),
+    ...(await planBuildRock(workingDir, id, imageOutputType)),
+    ...(await planBuildDockerImage(workingDir, id, imageOutputType))
+  ]
+}
+
+export async function run(): Promise<void> {
+  try {
+    let id = `${new Date().toISOString().replaceAll(':', '-').replace(/\..+/, '')}-${crypto.randomUUID().split('-')[3]}`
+    const identity = core.getInput('identifier')
+    if (identity.includes('__')) {
+      core.setFailed('identifier can not contain "__"')
```

Comment:
Out of curiosity, may I know the reason behind this?

Reply:
Because there are many restrictions in artifact names, so `__` is chosen as the separator in the artifact names. Restrict the use of `__` in user provided input to avoid ambiguity.